server/athenian/api/models/__init__.py

```diff
@@ -78,6 +80,48 @@ def create_base() -> BaseType:
     return declarative_base(cls=(Refreshable, Explodable))
 
 
+class SmartColumn(Column):
+    """Column that automatically converts `IN` and `NOT IN` to `ANY` and `!= ALL` repsecitvely"""
+
+    # 32767 is the maximum value for postgres:
+    #   - https://sentry.io/organizations/athenianco/issues/1785993425/
+    MAX_ALLOWED_QUERY_ARGUMENTS = 32767
+    # Anyway we cannot use exactly the upper limit because the length of the values used for
+    # `IN` and `NOT IN` could be near the the maximum allowed, but it often comes with other
+    # clauses that could make the total number of arguments bindings overflow the limit.
+    # This switch also affects performances:
+    #   - https://blog.jooq.org/2017/03/30/sql-in-predicate-with-in-list-or-with-array-which-is-faster/ # noqa: E501
+    IN_AUTO_ANY_THRESHOLD = 16383  # This specific value doesn't have any specific meaning
+
+    assert IN_AUTO_ANY_THRESHOLD < MAX_ALLOWED_QUERY_ARGUMENTS
+
+    def in_(self, other):
+        """Change the `in` operator into an `any` if too many values"""
+        log = logging.getLogger("smart_column.in_")
+        log.info("len(other) = %d", len(other))
+        if len(other) > self.IN_AUTO_ANY_THRESHOLD:
+            log.warning("switching to `any`")
+            # `in_` works if you pass a `dict` for example, but `any_` does not
+            return self == any_(other.__iter__())
+
+        log.info("doing nothing since len(other) = %d <= %d",
+                 len(other), self.IN_AUTO_ANY_THRESHOLD)
+        return super(SmartColumn, self).in_(other)
+
+    def notin_(self, other):
+        """Change the `notin` operator into a `!= all` if too many values"""
+        log = logging.getLogger("smart_column.notin_")
+        log.info("len(other) = %d", len(other))
+        if len(other) > self.IN_AUTO_ANY_THRESHOLD:
+            log.warning("switching to `!= all`")
+            # `notin_` works if you pass a `dict` for example, but `all_` does not
+            return self != all_(other.__iter__())
```

Comment:
I initially used `not_(self == any_(other))`, but it seems that `SQLAlchemy` is too smart... https://stackoverflow.com/questions/62952771/sqlalchemy-how-to-negate-an-any-expression